### PR TITLE
Use Frogpond's ccc-server to proxy and cache data

### PR DESCRIPTION
**Pull request by hawkrives**
aight

The St. Olaf routes are defined in https://github.com/frog-pond/ccc-server/blob/master/modules/node_modules/%40frogpond/ccci-stolaf-college/v1/news/index.mjs.

They're prefixed by `https://stolaf.api.frogpond.tech/v1`.

This server caches + proxies pretty much any data that the app uses. The exceptions are:

- streaming media, such as audio or video
- urls in content (we don't rewrite, say, BonApp's COR icon urls)
- private content (anything behind a user-based login wall; something that requires an app-level token, though, is fair game [eg, google calendar])

I think that's it.

---

I am not yet proxying the St. Olaf Multimedia streams list, although that's fair game; there's not yet an endpoint for that data in ccc-server (PRs welcome!), and I think we may want to stick a FilterToolbarThing (#2744) on it to allow filtering by `category` before/at the same time.

---

This also adds the ability for something that uses EventsList to swap out the "Detail" view; eg, the KSTO schedule could navigate to KSTOUpcomingEventDetail instead of EventDetailView and do something custom there. 

Additionally, this PR brings support for hiding news thumbnails on a per-source basis.

It also pulls out the AddToCalendar button from the EventDetailView into a reusable component thing, which I'm still liking. The component encapsulates the state machine of "adding to calendar", while letting the caller choose how to render the data.

All of these features came from CARLS, where I had originally done this ccc-server work.

---

<details><summary>Here's an example of the server log for me poking pretty much every endpoint in the app</summary>

```
rives@api ~> docker logs cccserver_stolaf_1 --follow --tail 100
  <-- GET /ping
  --> GET /ping 200 0ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 0ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 0ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 0ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 3ms 4b
  <-- GET /ping
  --> GET /ping 200 2ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 2ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /v1/food/named/menu/the-cage
  <-- GET /v1/food/named/cafe/the-cage
  --> GET /v1/food/named/menu/the-cage 200 23ms 13.46kb
  --> GET /v1/food/named/cafe/the-cage 200 6ms 1.93kb
  <-- GET /v1/food/named/menu/stav-hall
  <-- GET /v1/food/named/cafe/stav-hall
  --> GET /v1/food/named/cafe/stav-hall 200 424ms 2.14kb
  --> GET /v1/food/named/menu/stav-hall 200 640ms 11.03kb
  <-- GET /v1/spaces/hours
  --> GET /v1/spaces/hours 200 62ms 1.88kb
  <-- GET /v1/spaces/hours
  --> GET /v1/spaces/hours 304 2ms
  <-- GET /ping
  --> GET /ping 200 0ms 4b
  <-- GET /v1/news/named/stolaf
  --> GET /v1/news/named/stolaf 200 905ms 19.21kb
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /v1/news/named/stolaf
  --> GET /v1/news/named/stolaf 304 4ms
  <-- GET /v1/news/named/oleville
  --> GET /v1/news/named/oleville 200 1,227ms 2.95kb
  <-- GET /v1/news/named/mess
  --> GET /v1/news/named/mess 200 469ms 11.75kb
  <-- GET /v1/news/named/politicole
  --> GET /v1/news/named/politicole 200 301ms 513b
  <-- GET /v1/news/named/ksto
  --> GET /v1/news/named/ksto 200 514ms 11.54kb
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /v1/transit/bus
  --> GET /v1/transit/bus 200 39ms 1.95kb
  <-- GET /v1/dictionary
  --> GET /v1/dictionary 200 8ms 19.88kb
  <-- GET /v1/calendar/named/stolaf
  --> GET /v1/calendar/named/stolaf 200 267ms 2.95kb
  <-- GET /v1/calendar/named/stolaf
  --> GET /v1/calendar/named/stolaf 304 5ms
  <-- GET /v1/calendar/named/stolaf
  --> GET /v1/calendar/named/stolaf 304 3ms
  <-- GET /v1/calendar/named/oleville
  --> GET /v1/calendar/named/oleville 200 118ms 785b
  <-- GET /ping
  --> GET /ping 200 0ms 4b
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /v1/food/named/menu/stav-hall
  --> GET /v1/food/named/menu/stav-hall 304 15ms
  <-- GET /v1/food/named/cafe/stav-hall
  --> GET /v1/food/named/cafe/stav-hall 304 5ms
  <-- GET /ping
  --> GET /ping 200 1ms 4b
  <-- GET /v1/food/named/menu/the-cage
  --> GET /v1/food/named/menu/the-cage 304 12ms
  <-- GET /v1/food/named/cafe/the-cage
  --> GET /v1/food/named/cafe/the-cage 304 2ms
  <-- GET /ping
  --> GET /ping 200 4ms 4b
  <-- GET /v1/news/named/stolaf
  --> GET /v1/news/named/stolaf 304 7ms
  <-- GET /v1/news/named/ksto
  --> GET /v1/news/named/ksto 304 6ms
```

</details>

### CODE REVIEW

**Review by rye — CHANGES_REQUESTED** [review]
SGTM, but there are still a few stragglers in `source/lib/course-search/urls.js`:

```
$ git grep 'stodevx.github.io'
source/lib/course-search/urls.js:export const COURSE_DATA_PAGE = 'https://stodevx.github.io/course-data/'
source/lib/course-search/urls.js:export const INFO_PAGE = 'https://stodevx.github.io/course-data/info.json'
source/lib/course-search/urls.js:       'https://stodevx.github.io/course-data/data-lists/valid_gereqs.json'
source/lib/course-search/urls.js:       'https://stodevx.github.io/course-data/data-lists/valid_departments.json'
source/lib/course-search/urls.js:       'https://stodevx.github.io/course-data/data-lists/valid_times.json'
```

**Comment by hawkrives** [comment]
@rye Yeah, I'm wanting to leave those out of the initial PR, although not for the same reasons as Multimedia;

1. there's no endpoint on ccc-server for them yet
2. I don't know how I want to represent them on ccc-server yet (explorative PRs welcome)

so I'd like to leave course-search out of this PR, if that's acceptable.

**Comment by rye** [comment]
Well, since _all_ of the course data is separate from this repository, (which is the primary goal of switching away) I'm okay with this.

**Review by rye — DISMISSED** [review]

**Review by rye — APPROVED** [review]

File: source/views/menus/menu-bonapp.js [hunk 1]
```diff
@@ -83,7 +78,13 @@ export class BonAppHostedMenu extends React.PureComponent<Props, State> {
...
 
 	componentDidUpdate() {
-		if (this.state.cachedCafeId !== this.props.cafeId) {
+		let {cachedCafe} = this.state
+		let {cafe: newCafe} = this.props
+
+		cachedCafe = typeof cachedCafe === 'string' ? cachedCafe : cachedCafe.id
+		newCafe = typeof newCafe === 'string' ? newCafe : newCafe.id
+
+		if (cachedCafe !== newCafe) {
```
Comment (drewvolz) [inline source/views/menus/menu-bonapp.js:87]: 👍

File: source/views/menus/menu-bonapp.js [hunk 2]
```diff
@@ -94,19 +95,27 @@ export class BonAppHostedMenu extends React.PureComponent<Props, State> {
...
-		fetchJsonQuery(bonappMenuBaseUrl, {cafe: cafeId})
-	requestCafe = (cafeId: string) => () =>
-		fetchJsonQuery(bonappCafeBaseUrl, {cafe: cafeId})
-
 	fetchData = async (props: Props) => {
 		let cafeMenu: ?MenuInfoType = null
 		let cafeInfo: ?CafeInfoType = null
 
+		let menuUrl
+		let cafeUrl
```
Comment (drewvolz) [inline source/views/menus/menu-bonapp.js:103]: Can we please assign a default to these variables? `menuUrl` and `cafeUrl`

**Review by drewvolz — APPROVED** [review]
👍 :shipit:

File: source/views/menus/menu-bonapp.js (hunk 2 above)
Comment (hawkrives) [reply source/views/menus/menu-bonapp.js:103]: No. They throw an error if they're not assigned, though. 

Well, I suppose we could start with = null; would you like that?

Comment (hawkrives) [reply source/views/menus/menu-bonapp.js:103]: I do the same pattern in the calendar, fwiw

Comment (drewvolz) [reply source/views/menus/menu-bonapp.js:103]: Okay, it’s fine as-is.

Comment (hawkrives) [reply source/views/menus/menu-bonapp.js:103]: Side channel conversation with Drew has said I can merge without changing this.

**Pull request merged**